include/envoy/server/filter_config.h

```diff
@@ -0,0 +1,192 @@
+   */
+  virtual bool healthCheckFailed() PURE;
+
+  /**
+   * @return the server-wide http tracer.
+   */
+  virtual Tracing::HttpTracer& httpTracer() PURE;
+
+  /**
+   * @return the server's init manager. This can be used for extensions that need to initialize
+   *         after cluster manager init but before the server starts listening. All extensions
+   *         should register themselves during configuration load. initialize() will be called on
+   *         each registered target after cluster manager init but before the server starts
+   *         listening. Once all targets have initialized and invoked their callbacks, the server
+   *         will start listening.
+   */
+  virtual Init::Manager& initManager() PURE;
+
+  /**
+   * @return information about the local environment the server is running in.
+   */
+  virtual const LocalInfo::LocalInfo& localInfo() PURE;
+
+  /**
+   * @return RandomGenerator& the random generator for the server.
+   */
+  virtual Envoy::Runtime::RandomGenerator& random() PURE;
+
+  /**
+   * @return a new ratelimit client. The implementation depends on the configuration of the server.
+   */
+  virtual RateLimit::ClientPtr
+  rateLimitClient(const Optional<std::chrono::milliseconds>& timeout) PURE;
+
+  /**
+   * @return Runtime::Loader& the singleton runtime loader for the server.
+   */
+  virtual Envoy::Runtime::Loader& runtime() PURE;
+
+  /**
+   * DEPRECATED: Do not call this function. It will be removed post 1.4.0 and is needed for other
+   * deprecated functionality.
+   */
+  virtual Instance& server() PURE;
+
+  /**
+   * @return Stats::Scope& the filter's stats scope.
+   */
+  virtual Stats::Scope& scope() PURE;
+
+  /**
+   * @return ThreadLocal::Instance& the thread local storage engine for the server. This is used to
+   *         allow runtime lockless updates to configuration, etc. across multiple threads.
+   */
+  virtual ThreadLocal::Instance& threadLocal() PURE;
+};
+
+enum class NetworkFilterType { Read, Write, Both };
+
+/**
+ * This lambda is used to wrap the creation of a network filter chain for new connections as they
+ * come in. Filter factories create the lambda at configuration initialization time, and then they
+ * are used at runtime. This maps JSON -> runtime configuration.
+ */
+typedef std::function<void(Network::FilterManager& filter_manager)> NetworkFilterFactoryCb;
+
+/**
+ * Implemented by each network filter and registered via registerNamedNetworkFilterConfigFactory()
+ * or the convenience class RegisterNamedNetworkFilterConfigFactory.
+ */
+class NamedNetworkFilterConfigFactory {
+public:
+  virtual ~NamedNetworkFilterConfigFactory() {}
+
+  /**
+   * Create a particular network filter factory implementation.  If the implementation is unable to
+   * produce a factory with the provided parameters, it should throw an EnvoyException in the case
+   * of general error or a Json::Exception if the json configuration is erroneous.  The returned
```

Comment:
What if the factory lambda captures some JSON config and then finds an issue at runtime instantiation?